File: 8/lane_detection/src/edge_detection.py
import cv2
import numpy as np


def binary_array(array, thresh, value=0):

  if value == 0:
    binary = np.ones_like(array)   
  else:
    binary = np.zeros_like(array)  

  binary[(array >= thresh[0]) & (array <= thresh[1])] = value
 
  return binary
 

def blur_gaussian(image, ksize=3):

  return cv2.GaussianBlur(image, (ksize, ksize), 0)
         

def mag_thresh(image, sobel_kernel=3, thresh=(0, 255)):

  sobelx = np.absolute(sobel(image, orient='x', sobel_kernel=sobel_kernel))
  sobely = np.absolute(sobel(image, orient='y', sobel_kernel=sobel_kernel))
  mag = np.sqrt(sobelx ** 2 + sobely ** 2)
  
  return binary_array(mag, thresh)
 

def sobel(image, orient='x', sobel_kernel=3):
  
  if orient == 'x':
    sobel = cv2.Sobel(image, cv2.CV_64F, 1, 0, sobel_kernel)
  if orient == 'y':
    sobel = cv2.Sobel(image, cv2.CV_64F, 0, 1, sobel_kernel)
 
  return sobel
 

def threshold(image, thresh=(128,255), thresh_type=cv2.THRESH_BINARY):

  return cv2.threshold(image, thresh[0], thresh[1], thresh_type)

File: 8/lane_detection/src/edge_detection.cpp
#include "edge_detection.h"

cv::Mat binary_array(const cv::Mat& array, cv::Vec2f thresh, int value) {
    cv::Mat binary;

    if (value == 0) {
        binary = cv::Mat::ones(array.size(), array.type());
    } else {
        binary = cv::Mat::zeros(array.size(), array.type());
    }

    // Threshold aralığında kalan piksellerin maskesini oluştur
    cv::Mat mask;
    cv::inRange(array, thresh[0], thresh[1], mask);

    // mask = 255 olan yerleri value ile doldur
    binary.setTo(value, mask);

    return binary;
}

cv::Mat blur_gaussian(const cv::Mat& image, int ksize) {
    cv::Mat blurred_image;
    // Gaussian blur uygulama
    cv::GaussianBlur(image, blurred_image, cv::Size(ksize, ksize), 0);
    return blurred_image;
}

cv::Mat mag_thresh(const cv::Mat& image, int sobel_kernel, cv::Vec2f thresh) {
    // Apply Sobel in X and Y directions
    cv::Mat sobelx = cv::abs(sobel(image, "x", sobel_kernel));
    cv::Mat sobely = cv::abs(sobel(image, "y", sobel_kernel));

    // Compute magnitude
    cv::Mat mag;
    cv::magnitude(sobelx, sobely, mag);  // Calculate magnitude

    // Apply binary thresholding
    return binary_array(mag, thresh);
}

cv::Mat sobel(const cv::Mat& image, const std::string& orient, int sobel_kernel) {
    cv::Mat sobel_image;

    if (orient == "x") {
        // X yönü için Sobel filtre uygulama
        cv::Sobel(image, sobel_image, CV_64F, 1, 0, sobel_kernel);
    } else if (orient == "y") {
        // Y yönü için Sobel filtre uygulama
        cv::Sobel(image, sobel_image, CV_64F, 0, 1, sobel_kernel);
    }

    return sobel_image;
}

cv::Mat threshold(cv::Mat& image, cv::Scalar thresh, int thresh_type) {
    cv::Mat output;
    // Thresholding işlemi: image = giriş görüntüsü, thresh[0] = min, thresh[1] = max, thresh_type = eşik tipi
    cv::threshold(image, output, thresh[0], thresh[1], thresh_type);
    return output;
}

void printImageArray(const cv::Mat& image){
    for (int y = 0; y < image.rows; y++) {
        for (int x = 0; x < image.cols; x++) {
            uchar pixel_value = image.at<uchar>(y, x);
            std::cout << (int)pixel_value << " ";
        }
        std::cout << std::endl;
    }
}